File: hatch/hatch/report/resource_utilization/resource_utilization.py
import frappe
from frappe.utils import get_datetime


def execute(filters=None):
    filters = filters or {}

    columns = [
        {
            "label": "Resource",
            "fieldname": "resource",
            "fieldtype": "Link",
            "options": "Resource",
            "width": 150
        },
        {
            "label": "Total Bookings",
            "fieldname": "total_bookings",
            "fieldtype": "Int",
            "width": 120
        },
        {
            "label": "Total Hours Booked",
            "fieldname": "total_hours",
            "fieldtype": "Float",
            "width": 130
        },
        {
            "label": "Utilization %",
            "fieldname": "utilization",
            "fieldtype": "Percent",
            "width": 110
        },
        {
            "label": "Revenue",
            "fieldname": "revenue",
            "fieldtype": "Currency",
            "width": 120
        }
    ]

    conditions = {
        "booking_date": [
            "between",
            [filters.get("from_date"), filters.get("to_date")]
        ]
    }

    if filters.get("resource"):
        conditions["resource"] = filters["resource"]

    bookings = frappe.get_list(
        "Booking",
        filters=conditions,
        fields=[
            "resource",
            "booking_date",
            "start_time",
            "end_time",
            "total_amount"
        ]
    )

    data = {}

    for booking in bookings:
        resource = booking.resource

        if resource not in data:
            data[resource] = {
                "total_bookings": 0,
                "total_hours": 0,
                "revenue": 0
            }

        data[resource]["total_bookings"] += 1

        start = get_datetime(booking.start_time)
        end = get_datetime(booking.end_time)

        hours = (end - start).total_seconds() / 3600

        data[resource]["total_hours"] += hours
        data[resource]["revenue"] += booking.total_amount or 0

    rows = []

    period_hours = (
        get_datetime(filters["to_date"])
        - get_datetime(filters["from_date"])
    ).days + 1

    period_hours *= 24

    for resource, values in data.items():

        utilization = (
            values["total_hours"] / period_hours
            if period_hours
            else 0
        )

        rows.append({
            "resource": resource,
            "total_bookings": values["total_bookings"],
            "total_hours": values["total_hours"],
            "utilization": utilization,
            "revenue": values["revenue"]
        })

    total_bookings = sum(
        row["total_bookings"] for row in rows
    )

    total_revenue = sum(
        row["revenue"] for row in rows
    )

    busiest_resource = (
        max(
            rows,
            key=lambda row: row["total_bookings"]
        )["resource"]
        if rows
        else "-"
    )

    report_summary = [
        {
            "label": "Total Bookings",
            "value": total_bookings,
            "datatype": "Int"
        },
        {
            "label": "Total Revenue",
            "value": total_revenue,
            "datatype": "Currency"
        },
        {
            "label": "Busiest Resource",
            "value": busiest_resource,
            "datatype": "Data"
        }
    ]

    return columns, rows, None, report_summary

File: hatch/hatch/report/resource_utilization/resource_utilization.js
frappe.query_reports["Resource Utilization"] = {

    filters: [
        {
            fieldname: "from_date",
            label: "From Date",
            fieldtype: "Date",
            default: frappe.datetime.month_start(),
            reqd: 1
        },
        {
            fieldname: "to_date",
            label: "To Date",
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1
        },
        {
            fieldname: "resource",
            label: "Resource",
            fieldtype: "Link",
            options: "Resource"
        }
    ],

    formatter: function(value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (column.fieldname === "utilization") {

            let percentage = parseFloat(data.utilization) * 100;

            if (percentage > 95) {
                return `<span style="color:red;font-weight:bold">
                    ${percentage.toFixed(2)}%
                </span>`;
            }

            if (percentage >= 40 && percentage <= 85) {
                return `<span style="color:green;font-weight:bold">
                    ${percentage.toFixed(2)}%
                </span>`;
            }

            return `${percentage.toFixed(2)}%`;
        }

        return value;
    }
};